lib/datadog/security/reactive/operation.rb

```diff
@@ -0,0 +1,55 @@
+require 'datadog/security'
+require 'datadog/security/reactive/engine'
+
+module Datadog
+  module Security
+    module Reactive
+      class Operation
+        attr_reader :reactive
+        attr_reader :parent
+        attr_reader :name
+
+        # TODO: logger
+        def logger
+          @logger ||= ::Logger.new(STDOUT)
+          @logger.level = ::Logger::DEBUG
+          @logger
+        end
+
+        def initialize(name, parent = nil, reactive_engine = nil)
```

Comment:
I'm curious -- could you provide some info on how this class fits into the design? It seems mostly to be a) a reference to the current engine and b) something that manages the `Thread.current` variable, so I don't quite understand why we would want to create a new instance for every "operation", but that's probably because I'm not quite grokking how some of these pieces fit together.

Reply:
Operations represent a level of abstraction of something we want to track security for. Think of it as a `Span`/`SpanOperation` (the original name was actually Span). `Operation.new` is a bit like `Tracing.trace` which creates a new `Span`, but since there's no `Tracer` concept it hides tracking behind the scenes, so you just (conceptually, it can be spread across multiple nested method calls thanks to the thread variable) do:

```
 # block version, auto finalizes
Operation.new('root') do |op|
  ...
  op.publish('foo', foo)
  begin
    # non-block version, automatically sets the parent from thread context
    nested_op = Operation.new('nested')
    ...
    nested_op.publish('bar', bar)
  ensure
    nested_op.finalize
  end
  ...
end
```

Reactive engines can be overridden per operation, so that you can have different set of addresses to respond to. A feature not in this implementation is that you could add new addresses only available starting at a certain "operation scope" but still be able to react to original addresses:

```
 # block version, auto finalizes
Operation.new('root') do |op|
  op.subscribe('foo') { ... }
  
  # use a brand new engine, based off the parent one
  Operation.new('nested', reactive_engine: op.reactive.dup) do |nested_op|
    nested_op.subscribe('bar') { ... }
    
    Operation.new('nested2') do |deep_op|
      deep_op.publish('bar', bar) # works
      deep_op.publish('foo', foo) # works too, "inherited"
    end
  end
    
  op.publish('bar', bar) # 'bar' doesn't exist here
end
```

This (and the Reactive Engine) is a simpler version of the [Operation Stack](https://github.com/DataDog/datadog-instrumentation-gateway-ruby/blob/cb543f938b502de71dcdf9f1720c0bc8f16f10c5/opstack/op.rb).

Reply:
Can I convince you to add your comment as an actual comment on that file? ;)